### ISSUE TITLE
setPreSendingHandler seems to be not blocking and sometimes our attachment file is missing

### ISSUE THREAD

**Issue by fangpenlin** [Instabug/Instabug-React-Native#131]
To workaround the `addFileAttachment` missing bug https://github.com/Instabug/instabug-reactnative/issues/130

We use the `addFileAttachment` function from native module directly in `setPreSendingHandler` callback like this

```js
    const InstabugNative = NativeModules.Instabug
    Instabug.setPreSendingHandler(() => {
      const path = fs.DocumentDirectoryPath + '/log.txt'
      fs.exists(path).then((exists: boolean) => {
        if (exists) {
          InstabugNative.addFileAttachment(path)
        }
      })
    })
```

It works mostly, however we notice that sometimes the log file is not attached. I looked into the callback code

https://github.com/Instabug/instabug-reactnative/blob/5bfd6fab7d2b0b4da3e97d85af9c28877191df8b/ios/RNInstabug/InstabugReactBridge.m#L89-L97

and I wonder how this `setPreSendingHandler` works? Is it blocking callback, the report submits wait until this callback returns and then really submit? Or does the report submits in async manner?

Let's assume `setPreSendingHandler` callback is blocking submit, our native iOS code for attaching file to bug report like this

```
Instabug background thread
1. Prepare submit
2. Call setPreSendingHandler
3.  + hander attach log file
4. Submit the report
```

But for this React Native module, as the code in setPreSendingHandler is just

```
[self sendEventWithName:@"IBGpreSendingHandler" body:nil];
```

looks like it emits an `IBGpreSendingHandler` event in JavaScript thread, and as this `sendEventWithName` function call returns immediately, I guess there is a chance the report is submitted already, and it's too late for the hander in JavaScript to attach file for the report. Something like this may happen

```
Instabug background thread                              JavaScript thread
1. Prepare submit
2. Call setPreSendingHandler
3.  + sendEventWithName IBGpreSendingHandler
4. Submit the report
                                                        5. Attach log file
```

And if we are lucky, the attach log file happens before submit

```
Instabug background thread                              JavaScript thread
1. Prepare submit
2. Call setPreSendingHandler
3.  + sendEventWithName IBGpreSendingHandler
                                                        4. Attach log file
5. Submit the report
```

Can youguys confirm this is how it works? And if this is correct, is there any way we can ensure the file is attached to report?

**Comment by Korazy** [comment]
Hi @fangpenlin, thanks for all the details. Let me check this and get back to you.

**Comment by Korazy** [comment]
So here are the details, @fangpenlin:
1- preSendingHandler will block sending report so any fileAttachment added will be send with the report
2- We have a limitation on file attachments with maximum size being 5 mb as you can see in the documentation. We also log to the console that the file size is exceeding the limit and will not be attached with the bug.
3- We are checking if the path and file exist, if so we will attach the file with the report.

So the cases that will prevent the file from being attached with the report is if it’s size is exceeding 5 mb or if the file does not exist. Please let me know if that makes sense.

Regarding `[self sendEventWithName:@“IBGpreSendingHandler” body:nil];`, are you sure that it is a sync call? The reason I ask is because if it’s an async call we will not be able to guarantee the attaching of the file successfully.

**Comment by fangpenlin** [comment]
Hi @Korazy, `[self sendEventWithName:@"IBGpreSendingHandler" body:nil];` is async, as I said it returns immediately after queuing the event to be processed in the javascript thread, which implies it's async.

We're pretty sure the file size is less than 5MB and the file exists.

And yes, as `[self sendEventWithName:@"IBGpreSendingHandler" body:nil];` is async, the callback I passed to `setPreSendingHandler` in JavaScript won't guaranteed to be executed before submit the bug report, then the question is how can we workaround this if we need to attach file in the bug report?

Or put it this way, we had `setPreSendingHandler` in Objective-C code like this

```
    [Instabug setPreSendingHandler:^{
        [Instabug clearFileAttachments];
        NSArray<NSString *> *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *documentDir = paths[0];
        NSString *logFile = [documentDir stringByAppendingPathComponent:@"log.txt"];
        if ([[NSFileManager defaultManager] fileExistsAtPath:logFile] ) {
            [Instabug addFileAttachmentWithURL:[NSURL fileURLWithPath:logFile]];
        }
    }];
```

but the same code for React Native in js won't guarantee the file attachment happens before submit report, how can we fix this?

**Comment by Korazy** [comment]
The issue is that there is no way for us to know if the async call completed. So the only workaround would be to rely on sync calls.

**Comment by fangpenlin** [comment]
@Korazy To workaround this bug, I created a native React Native module using semaphore for blocking the `setPreSendingHandler` until `dispatch_semaphore_signal` is called. code like this

```obj-c

#import "RCTInstabugUtils.h"
#import <Instabug/Instabug.h>

NSString * const kPreSendingHandlerEvent = @"PreSendingHandlerEvent";

@implementation RCTInstabugUtils {
    dispatch_semaphore_t semaphore;
    int64_t timeout;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        semaphore = dispatch_semaphore_create(0);
        // by default, we wait 5 seconds until the event completes
        timeout = NSEC_PER_SEC * 5;
    }
    return self;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[kPreSendingHandlerEvent];
}

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue {
    return dispatch_get_main_queue();
}

RCT_EXPORT_METHOD(setBlockingPreSendingHandler:(BOOL)callBack) {
    if (!callBack) {
        [Instabug setPreSendingHandler:nil];
        return;
    }
    __weak __typeof(self) weakSelf = self;
    [Instabug setPreSendingHandler:^{
        if (!weakSelf) {
            return;
        }
        //
        [weakSelf sendEventWithName:kPreSendingHandlerEvent body:nil];
        // Wait for finishBlockingPreSending called or timeout, so that we
        // know the JavaScript thread has already did its job attaching
        // log file and do all other necessary stuff for submitting the
        // bug report.
        dispatch_time_t time;
        if (timeout == 0) {
            time = DISPATCH_TIME_NOW;
        } else if (timeout < 0) {
            time = DISPATCH_TIME_FOREVER;
        } else {
            time = dispatch_time(DISPATCH_TIME_NOW, timeout);
        }
        dispatch_semaphore_wait(semaphore, time);
    }];
}

RCT_EXPORT_METHOD(finishBlockingPreSending) {
    dispatch_semaphore_signal(semaphore);
}

RCT_EXPORT_METHOD(setTimeout:(int64_t)newTimeout) {
    timeout = newTimeout;
}

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@end
```

and the javascript code

```typescript
import * as Rx from 'rxjs/Rx'
import {
  EmitterSubscription,
  NativeEventEmitter,
  NativeModules,
} from 'react-native'

const { InstabugUtils } = NativeModules as any

interface BlockingPreSendingHandler {
  (): Rx.Observable<void>
}

let subscription: EmitterSubscription | null = null

export const setBlockingPreSendingHandler = (
  handler: BlockingPreSendingHandler | null
) => {
  if (subscription !== null) {
    subscription.remove()
    subscription = null
  }
  if (handler === null) {
    InstabugUtils.setBlockingPreSendingHandler(false)
    return
  }
  const emitter = new NativeEventEmitter(InstabugUtils)
  subscription = emitter.addListener('PreSendingHandlerEvent', () => {
    handler().subscribe(
      undefined,
      // when the observer returned from handler completes or errored, let's
      // finish blocking the pre sending handler
      () => {
        InstabugUtils.finishBlockingPreSending()
      },
      () => {
        InstabugUtils.finishBlockingPreSending()
      }
    )
  })
  InstabugUtils.setBlockingPreSendingHandler(true)
}
```

And the usage:

```typescript
    // Notice: Instabug they forgot to add addFileAttachment in index.js
    // of their exporting module, so that we need to get the native module
    // here in order to use addFileAttachment method. Created an issue here
    // ref: https://github.com/Instabug/instabug-reactnative/issues/130
    const InstabugNative = NativeModules.Instabug
    setBlockingPreSendingHandler(() => {
      const path = fs.DocumentDirectoryPath + '/log.txt'
      return Rx.Observable.fromPromise(fs.exists(path))
        .do((exists: boolean) => {
          if (exists) {
            InstabugNative.addFileAttachment(path)
          }
        })
        .mapTo(undefined)
    })
```

But I ran into a problem, when I report via instabug with screenshot, after the `Thank you` alert shows up for a short while, the app crashes

![screen shot 2018-03-22 at 1 51 24 pm](https://user-images.githubusercontent.com/201615/37798140-15c5e102-2dd9-11e8-80d0-19e7c2b95a1e.png)

Looks like it's calling `dispatch_group_leave`, any idea why it's crashing like this?

**Comment by fangpenlin** [comment]
hmmm https://github.com/envoy/instabug-reactnative  with the same approach but with the code change right inside instabug-reactnative native module, and it seems crash is gone.

**Comment by Kmohamed** [comment]
Hello @fangpenlin 
could you send me the exception name and full stack trace please ?

**Comment by fangpenlin** [comment]
@Kmohamed sure

```
* thread #2, queue = 'NSManagedObjectContext 0x1078a2b30', stop reason = EXC_BREAKPOINT (code=1, subcode=0x102531854)
  * frame #0: 0x0000000102531854 libdispatch.dylib`dispatch_group_leave + 76
    frame #1: 0x00000001018cd0b0 InstabugCore`__73-[IBGUserStepsImagesLocalAttachment prepareAttachmentPathWithCompletion:]_block_invoke_2 + 44
    frame #2: 0x00000001018cce34 InstabugCore`-[IBGUserStepsImagesLocalAttachment compressFiles:withCompletion:] + 1160
    frame #3: 0x00000001018cd060 InstabugCore`__73-[IBGUserStepsImagesLocalAttachment prepareAttachmentPathWithCompletion:]_block_invoke + 124
    frame #4: 0x000000010186599c InstabugCore`-[IBGUserStepsLogger loadUserStepsWithContext:sessionId:completionHandler:] + 1088
    frame #5: 0x00000001018438a8 InstabugCore`__43-[IBGDALService readASync:completionBlock:]_block_invoke_3 + 36
    frame #6: 0x000000018527ec48 CoreData`developerSubmittedBlockToNSManagedObjectContextPerform + 156
    frame #7: 0x000000010252d28c libdispatch.dylib`_dispatch_client_callout + 16
    frame #8: 0x000000010253abc8 libdispatch.dylib`_dispatch_queue_barrier_sync_invoke_and_complete + 120
    frame #9: 0x000000018527eb4c CoreData`-[NSManagedObjectContext performBlockAndWait:] + 304
    frame #10: 0x0000000101843830 InstabugCore`__43-[IBGDALService readASync:completionBlock:]_block_invoke_2 + 236
    frame #11: 0x000000010252d2cc libdispatch.dylib`_dispatch_call_block_and_release + 24
    frame #12: 0x000000010252d28c libdispatch.dylib`_dispatch_client_callout + 16
    frame #13: 0x000000010253bf80 libdispatch.dylib`_dispatch_queue_serial_drain + 696
    frame #14: 0x00000001025307ec libdispatch.dylib`_dispatch_queue_invoke + 332
    frame #15: 0x000000010253cf6c libdispatch.dylib`_dispatch_root_queue_drain_deferred_wlh + 428
    frame #16: 0x0000000102544020 libdispatch.dylib`_dispatch_workloop_worker_thread + 652
    frame #17: 0x0000000182642f1c libsystem_pthread.dylib`_pthread_wqthread + 932
    frame #18: 0x0000000182642b6c libsystem_pthread.dylib`start_wqthread + 4
```

**Comment by fangpenlin** [comment]
as my fork https://github.com/envoy/instabug-reactnative is working fine and it's doing exactly the same thing, so I guess maybe Instabug-iOS sdk is doing some initialization in terms of framework level. So when I create the native module inside my React Native app project directly, maybe somehow it creates another instance of instabug or some global initialization stuff. but anyway, just blind guessing as I can't read the source code.

**Comment by Kmohamed** [comment]
@fangpenlin I don’t have experience with react but initializing Instabug twice will make a lot of trouble anyway let me check with cross platform team and get back to you

**Comment by HeshamMegid** [comment]
A quick update on this: we have a fix for this issue that we'll be including in our next release.

**Comment by nasmuris** [comment]
@HeshamMegid was this already fixed? I'm using v8.0.23 and the attachement is not available in release on iOS

**Comment by HeshamMegid** [comment]
Unfortunately, we haven't shipped this yet. Will let you know once it's available.

**Comment by Buthrakaur** [comment]
Hello @HeshamMegid , any progress with this issue since February, please?

**Comment by HeshamMegid** [comment]
@Buthrakaur, unfortunately, this is pending on changes in our native SDKs that we haven't shipped yet. Will let you know once it's ready.

**Comment by HeshamMegid** [comment]
@Buthrakaur @nasmuris @fangpenlin We just shipped a fix for this in v8.3. Sorry this took a while from our side.

Please open a new issue if you have feedback about it.
